### PR TITLE
Switch ECMA_DIRECT_TYPE_INTEGER_VALUE and ECMA_DIRECT_TYPE_SIMPLE_VALUE

### PR DESCRIPTION
**Pull request by lygstate**
Reference to https://github.com/jerryscript-project/jerryscript/issues/4578

Signed-off-by: Yonggang Luo <luoyonggang@gmail.com>

### CODE REVIEW

**Review by rerobika — CHANGES_REQUESTED** [review]
That's possibly the IoT.js NAPI's implementation fault, how they represent a NAPI value. What I'm really cannot to understand that why are you trying to fix in in ECMA layer. This problem **must be** independent from the engine internals.

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 1]
```diff
@@ -277,10 +277,11 @@ extern inline bool JERRY_ATTR_CONST JERRY_ATTR_ALWAYS_INLINE
...
                                  ecma_value_t second_value) /**< second ecma value */
 {
-  JERRY_STATIC_ASSERT (ECMA_DIRECT_TYPE_INTEGER_VALUE == 0,
-                       ecma_direct_type_integer_value_must_be_zero);
+  JERRY_STATIC_ASSERT (ECMA_DIRECT_TYPE_INTEGER_VALUE == 8,
+                       ecma_direct_type_integer_value_must_be_8);
 
-  return ((first_value | second_value) & ECMA_DIRECT_TYPE_MASK) == ECMA_DIRECT_TYPE_INTEGER_VALUE;
+  return ((first_value & ECMA_DIRECT_TYPE_MASK) == ECMA_DIRECT_TYPE_INTEGER_VALUE)
+          && ((second_value & ECMA_DIRECT_TYPE_MASK) == ECMA_DIRECT_TYPE_INTEGER_VALUE);
```
[the remark this answers is not in the document]
Comment (lygstate) [reply jerry-core/ecma/base/ecma-helpers-value.c:284]: How benchmark it?

**Comment by lygstate** [comment]
> That's possibly the IoT.js NAPI's implementation fault, how they represent a NAPI value. What I'm really cannot to understand that why are you trying to fix in in ECMA layer. This problem **must be** independent from the engine internals.

Yeap, I've tried fixed it NAPI layer at the first place, but that's leading too much changes

**Comment by zherczeg** [comment]
I recommend try things internally before opening a PR. Running the tests reveal that arithmetic optimizations rely heavily on the zeroes, e.g. no need post-processing the a+b or a-b if the last bits are zero regardless of sign.

**Comment by lygstate** [comment]
> I recommend try things internally before opening a PR. Running the tests reveal that arithmetic optimizations rely heavily on the zeroes, e.g. no need post-processing the a+b or a-b if the last bits are zero regardless of sign.

First of all, I am new to jerryscript, so I always need help from the community, otherwise it's too much time consuming task.That's why I created this PR incase wrong direction.
Also, in the CI, I didn't found the performance testing related task, where are they?

**Comment by lygstate** [comment]
> I recommend try things internally before opening a PR. Running the tests reveal that arithmetic optimizations rely heavily on the zeroes, e.g. no need post-processing the a+b or a-b if the last bits are zero regardless of sign.

Where the a+b optimize you refering to? I only fond a^b a+b !a a*b need the tricky

**Comment by rerobika** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10992 -> 10992 : 0.000% | 13800 -> 13800 : 0.000% | 1.140 -> 1.143 : -0.284% | 
3d-morph.js | 289232 -> 289232 : 0.000% | 1576 -> 1576 : 0.000% | 1.340 -> 1.374 : -2.608% | 
3d-raytrace.js | 117656 -> 117656 : 0.000% | 2540 -> 2540 : 0.000% | 1.333 -> 1.334 : -0.124% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3432 -> 3432 : 0.000% | 0.776 -> 0.782 : -0.730% | 
access-fannkuch.js | 1792 -> 1792 : 0.000% | 1648 -> 1648 : 0.000% | 2.573 -> 2.608 : -1.366% | 
access-nbody.js | 4600 -> 4600 : 0.000% | 1736 -> 1736 : 0.000% | 1.530 -> 1.518 : +0.817% | 
bitops-3bit-bits-in-byte.js | 1024 -> 1024 : 0.000% | 1592 -> 1592 : 0.000% | 0.715 -> 0.756 : -5.722% | 
bitops-bits-in-byte.js | 984 -> 984 : 0.000% | 1592 -> 1592 : 0.000% | 1.010 -> 1.027 : -1.683% | 
bitops-bitwise-and.js | 752 -> 752 : 0.000% | 1344 -> 1344 : 0.000% | 1.226 -> 1.241 : -1.197% | 
bitops-nsieve-bits.js | 61184 -> 61184 : 0.000% | 1592 -> 1592 : 0.000% | 1.502 -> 1.525 : -1.579% | 
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 61584 -> 61584 : 0.000% | 0.529 -> 0.538 : -1.822% | 
crypto-aes.js | 28040 -> 28040 : 0.000% | 2288 -> 2288 : 0.000% | 0.880 -> 0.886 : -0.719% | 
crypto-md5.js | 70968 -> 70968 : 0.000% | 1704 -> 1704 : 0.000% | 0.745 -> 0.756 : -1.497% | 
crypto-sha1.js | 46328 -> 46328 : 0.000% | 1720 -> 1720 : 0.000% | 0.762 -> 0.769 : -0.874% | 
crypto.js | 51704 -> 51704 : 0.000% | 3156 -> 3156 : 0.000% | 8.720 -> 8.788 : -0.779% | 
date-format-tofte.js | 8952 -> 8952 : 0.000% | 2008 -> 2008 : 0.000% | 1.217 -> 1.215 : +0.184% | 
date-format-xparb.js | 15144 -> 15144 : 0.000% | 2552 -> 2552 : 0.000% | 1.638 -> 1.636 : +0.124% | 
deltablue.js | 455544 -> 455544 : 0.000% | 3540 -> 3540 : 0.000% | 5.435 -> 5.427 : +0.144% | 
math-cordic.js | 2048 -> 2048 : 0.000% | 1592 -> 1592 : 0.000% | 1.716 -> 1.727 : -0.665% | 
math-partial-sums.js | 1512 -> 1512 : 0.000% | 1568 -> 1568 : 0.000% | 1.149 -> 1.141 : +0.731% | 
math-spectral-norm.js | 3264 -> 3264 : 0.000% | 1592 -> 1592 : 0.000% | 0.741 -> 0.728 : **+1.828%** | 
raytrace.js | 22424 -> 22424 : 0.000% | 5000 -> 5000 : 0.000% | 2.645 -> 2.622 : +0.853% | 
richards.js | 8568 -> 8568 : 0.000% | 2140 -> 2140 : 0.000% | 0.310 -> 0.309 : +0.414% | 
string-base64.js | 90104 -> 90104 : 0.000% | 1640 -> 1640 : 0.000% | 1.411 -> 1.411 : +0.003% | 
string-fasta.js | 4400 -> 4400 : 0.000% | 1576 -> 1576 : 0.000% | 3.462 -> 3.428 : +0.976% | 
Geometric mean: | 0.000% | 0.000% | -0.613% | 

Binary (bytes) | master(b1f73e698a) | patch(0d9fea4620) | Diff |
----: | ----: | ----: | ----: |
size | 226976 | 226976 | 0 bytes |
.rodata | 20095 | 20095 | 0 bytes |
.dynstr | 446 | 446 | 0 bytes |
.rel.plt | 456 | 456 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 1008 | 1008 | 0 bytes |
.gnu.hash | 464 | 464 | 0 bytes |
.text | 200072 | 200352 | +280 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 244 | 244 | 0 bytes |
.plt | 716 | 716 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 408 | 408 | 0 bytes |
.gnu.version | 126 | 126 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

These are the benchmark results. Shows significant slowdown due to the extra branch in `ecma_are_values_integer_numbers`

**Comment by lygstate** [comment]
> +

No problem, that branch can be optimized, How to run the bench locally?

File: jerry-core/ecma/base/ecma-helpers-value.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (lygstate) [reply jerry-core/ecma/base/ecma-helpers-value.c]: Hi I've changed to bitwise operation, it's should be faster now:) May you have time to benchmark it again?

**Comment by rerobika** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10992 -> 10992 : 0.000% | 13800 -> 13800 : 0.000% | 1.144 -> 1.157 : -1.188% | 
3d-morph.js | 289232 -> 289232 : 0.000% | 1576 -> 1576 : 0.000% | 1.337 -> 1.373 : -2.704% | 
3d-raytrace.js | 117656 -> 117656 : 0.000% | 2540 -> 2540 : 0.000% | 1.336 -> 1.347 : -0.802% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3432 -> 3432 : 0.000% | 0.777 -> 0.784 : -0.883% | 
access-fannkuch.js | 1792 -> 1792 : 0.000% | 1648 -> 1648 : 0.000% | 2.564 -> 2.579 : -0.604% | 
access-nbody.js | 4600 -> 4600 : 0.000% | 1736 -> 1736 : 0.000% | 1.533 -> 1.534 : -0.082% | 
bitops-3bit-bits-in-byte.js | 1024 -> 1024 : 0.000% | 1592 -> 1592 : 0.000% | 0.716 -> 0.721 : -0.718% | 
bitops-bits-in-byte.js | 984 -> 984 : 0.000% | 1592 -> 1592 : 0.000% | 1.008 -> 1.018 : -1.017% | 
bitops-bitwise-and.js | 752 -> 752 : 0.000% | 1344 -> 1344 : 0.000% | 1.225 -> 1.216 : +0.797% | 
bitops-nsieve-bits.js | 61184 -> 61184 : 0.000% | 1592 -> 1592 : 0.000% | 1.498 -> 1.510 : -0.819% | 
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 61584 -> 61584 : 0.000% | 0.528 -> 0.535 : -1.299% | 
crypto-aes.js | 28040 -> 28040 : 0.000% | 2288 -> 2288 : 0.000% | 0.884 -> 0.883 : +0.084% | 
crypto-md5.js | 70968 -> 70968 : 0.000% | 1704 -> 1704 : 0.000% | 0.747 -> 0.750 : -0.463% | 
crypto-sha1.js | 46328 -> 46328 : 0.000% | 1720 -> 1720 : 0.000% | 0.761 -> 0.764 : -0.309% | 
crypto.js | 51704 -> 51704 : 0.000% | 3156 -> 3156 : 0.000% | 8.707 -> 8.778 : -0.821% | 
date-format-tofte.js | 8952 -> 8952 : 0.000% | 2008 -> 2008 : 0.000% | 1.214 -> 1.216 : -0.123% | 
date-format-xparb.js | 15144 -> 15144 : 0.000% | 2552 -> 2552 : 0.000% | 1.632 -> 1.615 : **+1.036%** | 
deltablue.js | 455544 -> 455544 : 0.000% | 3540 -> 3540 : 0.000% | 5.420 -> 5.459 : -0.718% | 
math-cordic.js | 2048 -> 2048 : 0.000% | 1592 -> 1592 : 0.000% | 1.716 -> 1.730 : -0.851% | 
math-partial-sums.js | 1512 -> 1512 : 0.000% | 1568 -> 1568 : 0.000% | 1.151 -> 1.150 : +0.142% | 
math-spectral-norm.js | 3264 -> 3264 : 0.000% | 1592 -> 1592 : 0.000% | 0.740 -> 0.751 : -1.421% | 
raytrace.js | 22424 -> 22424 : 0.000% | 5000 -> 5000 : 0.000% | 2.640 -> 2.636 : +0.171% | 
richards.js | 8568 -> 8568 : 0.000% | 2140 -> 2140 : 0.000% | 0.310 -> 0.310 : +0.107% | 
string-base64.js | 90104 -> 90104 : 0.000% | 1640 -> 1640 : 0.000% | 1.421 -> 1.413 : +0.550% | 
string-fasta.js | 4400 -> 4400 : 0.000% | 1576 -> 1576 : 0.000% | 3.452 -> 3.435 : +0.502% | 
Geometric mean: | 0.000% | 0.000% | -0.454% | 

Binary (bytes) | master(3df7998084) | patch(04e9705395) | Diff |
----: | ----: | ----: | ----: |
size | 226976 | 226976 | 0 bytes |
.rodata | 20095 | 20095 | 0 bytes |
.dynstr | 446 | 446 | 0 bytes |
.rel.plt | 456 | 456 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 1008 | 1008 | 0 bytes |
.gnu.hash | 464 | 464 | 0 bytes |
.text | 200040 | 200360 | +320 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 244 | 244 | 0 bytes |
.plt | 716 | 716 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 408 | 408 | 0 bytes |
.gnu.version | 126 | 126 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |


Still significant slowdown. Also I still suggest to fix it in IoT.js not in jerry.

My suggestion for https://github.com/jerryscript-project/iotjs/blob/7fdafb0e9937ae92f4d703387ee75c8264ee67ae/src/internal/node_api_internal.h#L29. 

```c
#define AS_JERRY_VALUE(nvalue) (jerry_value_t)(uintptr_t) (nvalue - 1)
#define AS_NAPI_VALUE(jval) (napi_value)(uintptr_t) (jval + 1)
```

**Comment by lygstate** [comment]
OK< then

**Pull request closed without merging**